auth/auth/auth.py

```diff
@@ -263,6 +341,36 @@ async def callback(request):
     return aiohttp.web.HTTPFound(next_page)
 
 
+@routes.post('/api/v1alpha/users/{user}/create')
+@rest_authenticated_developers_only
+async def create_user(request: web.Request, userdata):  # pylint: disable=unused-argument
+    db: Database = request.app['db']
+    username = request.match_info['user']
+
+    body = await request.json()
+    login_id = body['login_id']
+    is_developer = body['is_developer']
+    is_service_account = body['is_service_account']
+
+    try:
+        existing_user = await check_valid_new_user(db, username, login_id, is_developer, is_service_account)
+    except AuthUserError as e:
+        raise e.http_response()
+
+    if existing_user is not None:
+        return web.json_response()
+
+    await db.execute_insertone(
+        '''
+INSERT INTO users (state, username, login_id, is_developer, is_service_account)
+VALUES (%s, %s, %s, %s, %s);
+''',
+        ('creating', username, login_id, is_developer, is_service_account),
+    )
```

Comment:
This races with the select in `check_valid_new_user`. MySQL's [repeatable read](https://dev.mysql.com/doc/refman/8.0/en/innodb-consistent-read.html) isolation level does not guarantee that, if we a `SELECT` returns no rows that a subsequent `INSERT` will not cause a duplicate key error. See their example about `SELECT` and `UPDATE` where the update affects rows not seen by the select.

The `SELECT` in `users_with_username_or_login_id` needs a `LOCK IN SHARE MODE` to prevent other transactions from modifying/inserting the relevant rows.

I also think we should rework this so that both user creation functions use a common insertion function:

```python3
async def insert_new_user(tx, username, login_id, is_developer, is_service_account) -> Optional[dict]:
    existing_user = check_valid_new_user(tx, ...)
    if existing_user is not None:
        return existing_user
    db.execute_insertone(...)
    return None
```

The UI call should convert login_id to None before the call to `check_valid_new_user` anyway. I I think both the UI call and the API call can be implemented using `insert_new_user`.